接入SASLkafka集群，连接Jmx失败，后台报UnknownHostException

**kafka 集群配置如下：**
broker.id=1
listeners=SASL_PLAINTEXT://QL110:9092
advertised.listeners=SASL_PLAINTEXT://QL110:9092
num.network.threads=3
num.io.threads=8
socket.send.buffer.bytes=102400
socket.receive.buffer.bytes=102400
socket.request.max.bytes=104857600
log.dirs=/data/kafka-logs
num.partitions=1
num.recovery.threads.per.data.dir=1
offsets.topic.replication.factor=1
transaction.state.log.replication.factor=1
transaction.state.log.min.isr=1
default.replication.factor=3
log.cleanup.policy=delete
log.retention.hours=168
log.segment.bytes=1073741824
log.retention.check.interval.ms=300000
zookeeper.connect=10.91.198.2:2181,10.91.198.1:2181,10.91.198.3:2181
zookeeper.connection.timeout.ms=18000
group.initial.rebalance.delay.ms=0
sasl.enabled.mechanisms=SCRAM-SHA-256
sasl.mechanism.inter.broker.protocol=SCRAM-SHA-256
security.inter.broker.protocol=SASL_PLAINTEXT
authorizer.class.name=kafka.security.auth.SimpleAclAuthorizer
super.users=User:admin;User:tiuser

**brokers在zookeeper里的值为：**
get /brokers/ids/1
{"listener_security_protocol_map":{"SASL_PLAINTEXT":"SASL_PLAINTEXT"},"endpoints":["SASL_PLAINTEXT://QL110:9092"],"jmx_port":9999,"host":null,"timestamp":"1666333001145","port":-1,"version":4}

此时接入集群，正常填写zk、bootstrapServers、jmx、集群配置（SASL相关）配置，前端web response：
{"message":"成功","code":0,"data":{"jmxPort":9999,"kafkaVersion":null,"errList":[{"message":"连接Jmx失败","code":31,"data":null}],"zookeeper":"xxx.xxx.xxx.xx:2181"}}
如果此时忽略进行保存确定，后台会大量报：
2023-02-14 06:41:16.205 ERROR 48967 --- [-1-8-thread-484] c.x.k.s.k.c.s.p.i.PartitionServiceImpl   : method=batchGetPartitionOffsetFromKafkaAdminClient||clusterPhyId=1||topicName=OGGTOORACLE||offsetSpec=com.xiaojukeji.know.streaming.km.common.bean.entity.offset.KSOffsetSpec$KSLatestSpec@494c3a83||errMsg=exception!

java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TopicAuthorizationException: Topic authorization failed.
        at org.apache.kafka.common.internals.KafkaFutureImpl.wrapAndThrow(KafkaFutureImpl.java:45)
        at org.apache.kafka.common.internals.KafkaFutureImpl.access$000(KafkaFutureImpl.java:32)
        at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:89)
        at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:260)
        at com.xiaojukeji.know.streaming.km.core.service.partition.impl.PartitionServiceImpl.batchGetPartitionOffsetFromKafkaAdminClient(PartitionServiceImpl.java:393)
        at com.xiaojukeji.know.streaming.km.core.service.version.impl.VersionControlServiceImpl.doHandler(VersionControlServiceImpl.java:93)
        at com.xiaojukeji.know.streaming.km.core.service.version.BaseVersionControlService.doVCHandler(BaseVersionControlService.java:62)
        at com.xiaojukeji.know.streaming.km.core.service.version.BaseKafkaVersionControlService.doVCHandler(BaseKafkaVersionControlService.java:29)
        at com.xiaojukeji.know.streaming.km.core.service.partition.impl.PartitionServiceImpl.getAllPartitionOffsetFromKafka(PartitionServiceImpl.java:195)
        at com.xiaojukeji.know.streaming.km.core.service.cluster.impl.ClusterMetricServiceImpl.getMessageSize(ClusterMetricServiceImpl.java:352)
        at com.xiaojukeji.know.streaming.km.core.service.version.impl.VersionControlServiceImpl.doHandler(VersionControlServiceImpl.java:93)
        at com.xiaojukeji.know.streaming.km.core.service.version.BaseVersionControlService.doVCHandler(BaseVersionControlService.java:62)
        at com.xiaojukeji.know.streaming.km.core.service.version.BaseKafkaVersionControlService.doVCHandler(BaseKafkaVersionControlService.java:29)
        at com.xiaojukeji.know.streaming.km.core.service.cluster.impl.ClusterMetricServiceImpl.collectClusterMetricsFromKafka(ClusterMetricServiceImpl.java:208)
        at com.xiaojukeji.know.streaming.km.collector.metric.kafka.ClusterMetricCollector.lambda$collectKafkaMetrics$0(ClusterMetricCollector.java:57)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.kafka.common.errors.TopicAuthorizationException: Topic authorization failed.


**解决方法：**
com.xiaojukeji.know.streaming.km.persistence.kafka.zookeeper.service.impl.KafkaZKDAOImpl.class 里的 getBrokerMetadata(String zkAddress) 方法体（55-56行之间）里增加BrokerMetadata.parseAndUpdateBrokerMetadata(metadata);逻辑：如果host为空，就取endpoint里的值，具体详见方法体，然后重新打包部署

--- a/km-persistence/src/main/java/com/xiaojukeji/know/streaming/km/persistence/kafka/zookeeper/service/impl/KafkaZKDAOImpl.java
+++ b/km-persistence/src/main/java/com/xiaojukeji/know/streaming/km/persistence/kafka/zookeeper/service/impl/KafkaZKDAOImpl.java
@@ -7,6 +7,7 @@ import com.xiaojukeji.know.streaming.km.common.bean.entity.broker.Broker;
 import com.xiaojukeji.know.streaming.km.common.bean.entity.kafkacontroller.KafkaController;
 import com.xiaojukeji.know.streaming.km.common.bean.entity.result.ResultStatus;
 import com.xiaojukeji.know.streaming.km.common.bean.entity.topic.Topic;
+import com.xiaojukeji.know.streaming.km.common.constant.Constant;
 import com.xiaojukeji.know.streaming.km.common.enums.topic.TopicTypeEnum;
 import com.xiaojukeji.know.streaming.km.common.exception.AdminOperateException;
 import com.xiaojukeji.know.streaming.km.common.exception.NotExistException;
@@ -78,7 +79,7 @@ public class KafkaZKDAOImpl implements KafkaZKDAO {
 
         try {
             BrokerMetadata metadata = this.getData(kafkaZkClient.currentZooKeeper(), BrokerIdZNode.path(brokerId), false, BrokerMetadata.class);
-            BrokerMetadata.parseAndUpdateBrokerMetadata(metadata);
+
             return this.convert2Broker(clusterPhyId, brokerId, metadata);
         } catch (KeeperException ke) {
             logger.error("method=getBrokerMetadata||clusterPhyId={}||brokerId={}||errMsg=exception", clusterPhyId, brokerId, ke);
@@ -279,7 +280,7 @@ public class KafkaZKDAOImpl implements KafkaZKDAO {
         metadata.setJmxPort(brokerMetadata.getJmxPort());
         metadata.setStartTimestamp(brokerMetadata.getTimestamp());
         metadata.setRack(brokerMetadata.getRack());
-        metadata.setStatus(1);
+        metadata.setStatus(Constant.ALIVE);
         metadata.setEndpointMap(brokerMetadata.getEndpointMap());
         return metadata;
     }

--- a/km-persistence/src/main/java/com/xiaojukeji/know/streaming/km/persistence/kafka/zookeeper/znode/brokers/BrokerMetadata.java
+++ b/km-persistence/src/main/java/com/xiaojukeji/know/streaming/km/persistence/kafka/zookeeper/znode/brokers/BrokerMetadata.java
@@ -1,12 +1,11 @@
 package com.xiaojukeji.know.streaming.km.persistence.kafka.zookeeper.znode.brokers;
 
-import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.alibaba.fastjson.JSON;
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.xiaojukeji.know.streaming.km.common.bean.entity.common.IpPortData;
 import com.xiaojukeji.know.streaming.km.common.constant.KafkaConstant;
 import com.xiaojukeji.know.streaming.km.common.utils.ConvertUtil;
-import lombok.Data;
 
 import java.io.Serializable;
 import java.util.HashMap;
@@ -51,7 +50,6 @@ import java.util.Map;
  * }
  *
  */
-@Data
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class BrokerMetadata implements Serializable {
     private static final long serialVersionUID = 3918113492423375809L;
@@ -74,34 +72,92 @@ public class BrokerMetadata implements Serializable {
 
     private String rack;
 
-    @JsonIgnore
-    public String getExternalHost() {
-        if (!endpointMap.containsKey(KafkaConstant.EXTERNAL_KEY)) {
-            // external如果不存在，就返回host
-            return host;
-        }
-
-        return endpointMap.get(KafkaConstant.EXTERNAL_KEY).getIp();
+    public List<String> getEndpoints() {
+        return endpoints;
     }
 
-    @JsonIgnore
-    public String getInternalHost() {
-        if (!endpointMap.containsKey(KafkaConstant.INTERNAL_KEY)) {
-            // internal如果不存在，就返回host
-            return host;
-        }
-        return endpointMap.get(KafkaConstant.INTERNAL_KEY).getIp();
+    public void setEndpoints(List<String> endpoints) {
+        this.endpoints = endpoints;
     }
 
-    public static void parseAndUpdateBrokerMetadata(BrokerMetadata brokerMetadata) {
-        brokerMetadata.setEndpointMap(new HashMap<>());
+    public Map<String, IpPortData> getEndpointMap() {
+        if (endpointMap == null) {
+            this.parseBrokerMetadata();
+        }
 
-        if (brokerMetadata.getEndpoints().isEmpty()) {
+        return endpointMap;
+    }
+
+    public String getHost() {
+        if (endpointMap == null) {
+            this.parseBrokerMetadata();
+        }
+
+        return host;
+    }
+
+    public void setHost(String host) {
+        this.host = host;
+    }
+
+    public Integer getPort() {
+        if (endpointMap == null) {
+            this.parseBrokerMetadata();
+        }
+
+        return port;
+    }
+
+    public void setPort(Integer port) {
+        this.port = port;
+    }
+
+    public Integer getJmxPort() {
+        return jmxPort;
+    }
+
+    public void setJmxPort(Integer jmxPort) {
+        this.jmxPort = jmxPort;
+    }
+
+    public Integer getVersion() {
+        return version;
+    }
+
+    public void setVersion(Integer version) {
+        this.version = version;
+    }
+
+    public Long getTimestamp() {
+        return timestamp;
+    }
+
+    public void setTimestamp(Long timestamp) {
+        this.timestamp = timestamp;
+    }
+
+    public String getRack() {
+        return rack;
+    }
+
+    public void setRack(String rack) {
+        this.rack = rack;
+    }
+
+    private synchronized void parseBrokerMetadata() {
+        if (this.endpointMap != null) {
             return;
         }
 
+        if (this.endpoints == null || this.endpoints.isEmpty()) {
+            this.endpointMap = new HashMap<>(0);
+            return;
+        }
+
+        Map<String, IpPortData> tempEndpointMap = new HashMap<>();
+
         // example EXTERNAL://10.179.162.202:7092
-        for (String endpoint: brokerMetadata.getEndpoints()) {
+        for (String endpoint: this.endpoints) {
             int idx1 = endpoint.indexOf("://");
             int idx2 = endpoint.lastIndexOf(":");
             if (idx1 == -1 || idx2 == -1 || idx1 == idx2) {
@@ -111,19 +167,37 @@ public class BrokerMetadata implements Serializable {
             String brokerHost = endpoint.substring(idx1 + "://".length(), idx2);
             String brokerPort = endpoint.substring(idx2 + 1);
 
-            brokerMetadata.getEndpointMap().put(endpoint.substring(0, idx1), new IpPortData(brokerHost, brokerPort));
+            tempEndpointMap.put(endpoint.substring(0, idx1), new IpPortData(brokerHost, brokerPort));
 
             if (KafkaConstant.INTERNAL_KEY.equals(endpoint.substring(0, idx1))) {
                 // 优先使用internal的地址进行展示
-                brokerMetadata.setHost(brokerHost);
-                brokerMetadata.setPort(ConvertUtil.string2Integer(brokerPort));
+                this.host = brokerHost;
+                this.port = ConvertUtil.string2Integer(brokerPort);
             }
 
-            if (null == brokerMetadata.getHost()) {
-                brokerMetadata.setHost(brokerHost);
-                brokerMetadata.setPort(ConvertUtil.string2Integer(brokerPort));
+            if (null == this.host) {
+                this.host = brokerHost;
+                this.port = ConvertUtil.string2Integer(brokerPort);
             }
         }
+
+        this.endpointMap = tempEndpointMap;
+    }
+
+    public static void main(String[] args) {
+        String str = "{\t\n" +
+                "\t\"listener_security_protocol_map\":{\"EXTERNAL\":\"SASL_PLAINTEXT\",\"INTERNAL\":\"SASL_PLAINTEXT\"},\n" +
+                "\t\"endpoints\":[\"EXTERNAL://10.179.162.202:7092\",\"INTERNAL://10.179.162.202:7093\"],\n" +
+                "\t\"jmx_port\":8099,\n" +
+                "\t\"host\":null,\n" +
+                "\t\"timestamp\":\"1627289710439\",\n" +
+                "\t\"port\":-1,\n" +
+                "\t\"version\":4\n" +
+                "}";
+
+        BrokerMetadata bm = JSON.parseObject(str, BrokerMetadata.class);
+        System.out.println(bm.getHost());
+        System.out.println(JSON.toJSON(bm));
     }
 }
 

